File: DraftRightMobile/android/app/src/main/kotlin/com/draftright/keyboard/voice/MicHoldGesture.kt
package com.draftright.keyboard.voice

import kotlin.math.hypot

/**
 * Pure geometry for the hold-to-talk mic gesture. Kept android-free so it is
 * unit-testable without touch events. The view feeds displacement from the
 * touch-down point; this decides whether the finger has slid far enough to
 * mean "cancel" (slide-away-to-cancel, Zalo/WhatsApp style).
 */
object MicHoldGesture {
    /** Delay before a hold starts recording; a shorter press is a no-op tap. */
    const val ARM_MS = 180L
    /** Default slide-to-cancel threshold, in dp (the view converts to px). */
    const val DEFAULT_SLOP_DP = 32

    /** True once the finger has slid [slopPx] or more from the down point. */
    fun isCancelArmed(dxPx: Float, dyPx: Float, slopPx: Float): Boolean =
        hypot(dxPx, dyPx) >= slopPx
}

File: DraftRightMobile/ios/DraftRightKeyboardCore/Sources/DraftRightKeyboardCore/Voice/MicHoldGesture.swift
import CoreGraphics

/// Pure geometry for the hold-to-talk mic gesture. Kept UIKit-free so it is
/// unit-testable without touch events. The view feeds displacement from the
/// touch-down point; this decides whether the finger has slid far enough to
/// mean "cancel" (slide-away-to-cancel, Zalo/WhatsApp style). Mirrors the
/// Android `MicHoldGesture`.
public enum MicHoldGesture {
    /// Delay before a hold starts recording; a shorter press is a no-op tap.
    public static let armMs = 180

    /// Default slide-to-cancel threshold, in points.
    public static let defaultSlop: CGFloat = 32

    /// True once the finger has slid `slop` or more up or left from the down
    /// point. Screen y grows downward, so "up" is a negative dy and "left" is a
    /// negative dx. Deliberately NOT omnidirectional: a finger relaxing or the
    /// phone tilting down while speaking is common and shouldn't silently
    /// discard a good dictation, so downward/rightward drift never arms cancel.
    public static func isCancelArmed(dx: CGFloat, dy: CGFloat, slop: CGFloat) -> Bool {
        dy <= -slop || dx <= -slop
    }
}
